### DISCUSSION TITLE
Release planning

### DISCUSSION THREAD

**Discussion by xblax** [xblax/flashforge_ad5m_klipper_mod#99] [Development]
Starting a general thread about release planning.

I think many of the first public beta issues were fixed, another beta release is due. Does anyone see any issues/improvements that are already solved but did not make it into the master branch?

@consp and other GitHub collaborators:
I did already create a draft Release for the v00.04-beta version. Do you see anything that's missing?
If nobody disagrees I also think that KlipperScreen can be promoted from alpha to beta.

**Comment by KaruroChori** [comment]
I think the list for 0.04 is fine, but I have not been able myself updated over the last week, so I might be missing something. Finally I am back printing.

**Comment by consp** [comment]
> Starting a general thread about release planning.
> 
> I think many of the first public beta issues were fixed, another beta release is due. Does anyone see any issues/improvements that are already solved but did not make it into the master branch?
> 
> @consp and other GitHub collaborators: I did already create a draft Release for the v00.04-beta version. Do you see anything that's missing? If nobody disagrees I also think that KlipperScreen can be promoted from alpha to beta.

I'll have a look tonight and test it as well, currently only a few commits behind so shouldn't be too much of an issue.

KS works good enough. Still a bit slow, I'm considering looking into the lvgl idea with guppyscreen if it's more responsive since klipperscreen works and i'm not sure it can be more performant, should also use less memory. Biggest problem there is WiFi support. Lot easier to do in python than c++, but there are ways around it. That's something for another day though, for now the way to go is KlipperScreen as its easily modifiable and not too hard to run (and mostly stable except some quirks).

**Reply by consp** [reply]
Biggest point I'd make was the warning for the config files, but that's already there!

I still wonder why most people have not reported the wining noise of the servo, maybe it's just us unlucky us. Could also be it's not an issue with all models.

**Reply by KaruroChori** [reply]
I did an initial attempt to have guppyscreen compiled for this printer. I cannot recall all details since it was late and I was in a rush as I had to leave the day after, but I had to fight a lot with their makefiles to even get a compiled build done. And at the end, it started but got stuck during initialization.

I also have the demo environment configured with lvgl 9.0 and few more libraries which I reported a while back: https://github.com/KaruroChori/lvgl
Sadly I had to fix their makefile to work with meson, and they force pushed changes to my fork so I am not even sure it works now.

**Reply by xblax** [reply]
Since we have the ability to support different variants from the master branch, we could definitely add one for Guppy Screen as well. Maybe it could be compiled as Buildroot package instead of compiling it externally. No idea how well it works fro printers other than the Creality K1 Series.

**Reply by consp** [reply]
Well, it was easier then I thought (not clean though, needed to remove all references to wpa and it might need some aditional patching). Touch is also not working yet but that's probably xorg messing about. The cross compile also needs some patching to get libhv compiled as it should.

![IMG_20240405_100803_MP](https://github.com/xblax/flashforge_adm5_klipper_mod/assets/6802998/38d2be49-1a6d-4a4e-a993-ea7bfbfbf16f)


Lot lighter on the memory as well:
```
free -h
              total        used        free      shared  buff/cache   available
Mem:         107.8M       62.3M        6.3M       56.0K       39.3M       42.0M
Swap:        128.0M       19.5M      108.5M
```

**Reply by consp** [reply]
Hardcoded evdev for now (we do not want this, because cameras screw up the evdev order) and it is a LOT more responsive. Going to take some work to get it up to the same functionality as klipperscreen but at least I can home it. I guess all moonraker/klipper standard commands will work, things system specific (blanking, wifi etc) won't.

Also need to configure dpi/icons properly, since neither default settings are great.

**Reply by consp** [reply]
Is there a reason for not using wpa_supplicant (besides it being a pita) and going for iwd? Almost everything uses wpa_sup which would make porting a lot easier.

**Reply by xblax** [reply]
Nice - how do you build it? Manually or did you create a buildroot package? 

Reason for choosing iwd was that it's the "new" thing and supposedly more memory efficient due to relying mostly on kernel built-in features. Afaik wpa_supplicant in Buildroot also does not come with a nice daemon that does auto-(re)connect on system start but I might be mistaken. I did not have the GUI applications in mind, initially. I don't like solutions where enabling of Wifi depends on the GUI app running, like it is with the Stock Flashforge software.

NetworkManager has an IWD-backend, maybe that's worth looking at to see how IWD can be controlled from C++.
https://iwd.wiki.kernel.org/networkmanager But we can also give wpa_supplicant a try if it helps.

**Reply by consp** [reply]
> Manually

Currently yes, should be doable with a buildpkg but since I needed to figure out all the submodules first it was easier to not go for a buildroot package and make it work. First test is creating the binary, if it works I'll have a look at the package.

I'm now trying both iwd/wpa combined since the both use nl80211 it should cooperate (a bit at least). Let's see how it goes. The deamon is not too heavy by the looks of it. 

>  auto-(re)connect on system start 

If I'm not mistaken it will if auto-scan is enabled (which is a buildroot option). Will check after rebuild.

> IWD-backend

Looked at iwlib (which is available and pretty much the same level). But that's quite a rewrite as it contains all the nicities (not) of the kernel interface and lacks all the nicities of any commanline tooling. Bit more complicated than the python version 😆 


```
 PID User Command                                                                                                                                                                         Swap   USS   PSS   RSS
1238 root udhcpc -t1 -A3 -b -R -O search -O staticroutes -p /var/run/udhcpc.eth0.pid -i eth0 -x hostname:ADM5                                                                               84    12   125  1220
1270 root httpd -p 4000 -f -h /root/printer_software/web/mainsail                                                                                                                           84     8   146  1704
1266 root httpd -p 4001 -f -h /root/printer_software/web/fluidd                                                                                                                             84     4   149  1752
1095 root /sbin/syslogd -n                                                                                                                                                                  40    48   201  1864
1099 root /sbin/klogd -n                                                                                                                                                                    36    52   204  1808
1248 root /usr/sbin/dropbear -R                                                                                                                                                             24    64   242  1392
1244 root /usr/sbin/ntpd -n -S fake-hwclock -p pool.ntp.org                                                                                                                                 28    80   256  2064
1777 root /usr/bin/ts_uinput                                                                                                                                                                 0   184   257  1468
1111 root /sbin/mdev -df                                                                                                                                                                   112   224   410  2160
2373 root -sh                                                                                                                                                                                0   228   417  2072
1128 dbus dbus-daemon --system                                                                                                                                                              60   400   453  1380
   1 root init                                                                                                                                                                               4    76   464   852
1326 root /sbin/getty -L ttyS3 115200 vt100                                                                                                                                                  0   112   500   888
2372 root /usr/sbin/dropbear -R -2 9                                                                                                                                                         0   324   538  2020
1142 root /usr/libexec/iwd                                                                                                                                                                  24  1388  1499  2908
2090 root wpa_supplicant -B -Dnl80211 -iwlan0 -c/etc/wpa_supplicant.conf                                                                                                                     0  1920  2020  3104
2429 root ./guppyscreen                                                                                                                                                                      0  5804  6261  8064
1295 root ustreamer -d /dev/video0 -r 1280x720 -m MJPEG --device-timeout=2 -w 1 -I MMAP -c HW -s* -p 8080                                                                                    0  6200  6316  7796
2436 root /usr/bin/python3 /usr/bin/smem -a                                                                                                                                                  0  5508  6894 10880
1256 root /root/printer_software/klipper/klippy-env/bin/python3 /root/printer_software/klipper/klippy/klippy.py /root/printer_data/config/printer.cfg -l /root/printer_data/logs/klipper. 1076 14208 16485 21364
1262 root /root/printer_software/moonraker/moonraker-env/bin/python3 /root/printer_software/moonraker/moonraker/moonraker.py -d /root/printer_data                                         276 25664 27676 32396
```

Look like iwd and wpa_sup are about the same wrt memory usage.

Guppyscreen has some quirks though. Took me a while to find out how to enter the wifi password (you have to press the non-visible input field)

Do need to write a input event selection thing since it's hardcoded in the code at the moment and if someone adds a usb device with buttons (e.g. a webcam) it's not going to work properly.

Looks like guppyscreen is not on-par with klipperscreen feature wise but it runs a lot smoother and less bloaty.

**Reply by xblax** [reply]
It would also be possible to enable/disable IWD just in specific variants, if that helps. 

I could also very well accept a solution where we just remove Wifi settings from Guppyscreen completely and have the users create the IWD config file manually. 

I just have the IWD config file on my USB Stick. Created once and then never have to care about it again. It's not much different to setting up a Raspberry Pi for WiFi the first time and putting the WPA supplicant config on the micro SD card.

**Reply by consp** [reply]
> If I'm not mistaken it will if auto-scan is enabled (which is a buildroot option). Will check after rebuild.

It doesn't.

**Reply by consp** [reply]
> I could also very well accept a solution where we just remove Wifi settings from Guppyscreen completely and have the users create the IWD config file manually.

So far I've seen no adverse effects of having both iwd and wpa running. wpa connects (but needs dhcp enabled on the interface), iwd auto-connects. It's whatever people want to use in that case. I'm leaning towards also enabeling wpa on klipperscreen/guppyscreen to allow the original code to be used instead of monkey-patching a iwd variant into it. It would change nothing with respect to how iwd works and allow for wpa to also function.

~~Of course, wpa doesn't have data so there's that, which is probably the reason there is not auto-connect~~ ~~after updating the config of course it will save it but not autoconnect.~~ somehow it now auto connects after changing the autoscan settings 🤷

**Reply by consp** [reply]
Biggest issue I'm having so far is lack of prompt support, don;t know if that's simply not done yet or is broken.

**Reply by consp** [reply]
> Biggest issue I'm having so far is lack of prompt support, don;t know if that's simply not done yet or is broken.

Well, my hacked together c contraption wrapped around some c++ code seems to work so that might not be an issue anymore. I can shut down at least (options are VERY limited, colors do not work and that's because lvgl cannot do different button colors in a msgbox).

This fixes the "I cannot shut down when I'm at the printer" "problem" I was having.

Are there any other commands with prompts?

**Reply by KaruroChori** [reply]
What do you mean with "commands with prompts"?

**Reply by consp** [reply]
> commands with prompts

Anything which uses action:prompt_begin

I was under the idea the save_config also had this but that is not the case.

**Reply by ballaswag** [reply]
Take a look at this build configuration I use for the CM1 on a btt pad 7 https://github.com/ballaswag/guppyscreen/blob/bcfa539b941f616da25d4958f9eb15525901138b/.github/workflows/build.yml#L35

It sets up a few things for make to build for ARM, like cc and cxx, and some guppy specific flags like evcalibration, rotation, and target small size screens.

**Reply by consp** [reply]
> ARM, like cc and cxx,

Pretty much the way I went. Strange thing is the libhv cmake ignored the cross_compile but maybe that's an error on my side. Adding a CMAKE_CXX_COMPILER define and it works like a charm. 

> evcalibration

Fortunately it's not needed, the touchscreen is calibrated stock via tslib and with ts_uinput you can make a calibrated evdev.

**Comment by consp** [comment]
> Do you see anything that's missing?

Added #68 to the "servo issue" since the images are there. You made a pretty good coverage of all the fixed issues. Nothing jumps out as missing.

**Comment by KaruroChori** [comment]
For reference: https://github.com/xblax/flashforge_adm5_klipper_mod/discussions/57#discussioncomment-9026936  
If I recall correctly, consp was able to integrate the patched file at some point, but it was not carried over for the latest release.  
Should we get it back for the next one?

**Reply by xblax** [reply]
Yes, I'm planning to do that. Probably for the next release. I'd like to make Klipper, Moonraker and Mainsail Buildroot packages even if there is not much to compile, then we can use the automatic patch handling etc.
